From 861879edc67cdf244d1f7d1ea927f7c1a9399263 Mon Sep 17 00:00:00 2001
From: Mark Thomas <markt@apache.org>
Date: Thu, 3 Nov 2016 09:01:35 +0000
Subject: [PATCH] Fix DBCP-459

Ensure that a thread's interrupt status is visible to the caller if the
thread is interrupted during a call to PoolingDataSource.getConnection()
---
 src/changes/changes.xml                             | 13 +++++++++----
 .../org/apache/commons/dbcp2/PoolingDataSource.java |  4 ++++
 2 files changed, 13 insertions(+), 4 deletions(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index da71653d37..538819d75c 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -109,10 +109,15 @@ The <action> type attribute can be add,update,fix,remove.
         remains but is deprecated so not to break clients currently using the
         incorrect name.
       </action>
-     <action dev="markt" type="add" issue="DBCP-462" due-to=" Keiichi Fujino">
-       Refactoring to prepare for a future patch to enable pooling of all
-       prepared and callable statements in PoolingConnection.
-     </action>
+       <action dev="markt" type="add" issue="DBCP-462" due-to=" Keiichi Fujino">
+         Refactoring to prepare for a future patch to enable pooling of all
+         prepared and callable statements in PoolingConnection.
+       </action>
+       <action dev="markt" type="fix" issue="DBCP-459">
+         Ensure that a thread's interrupt status is visible to the caller if the
+         thread is interrupted during a call to
+         PoolingDataSource.getConnection().
+       </action>
     </release>
     <release version="2.1.1" date="6 Aug 2015" description=
 "This is a patch release, including bug fixes only.">
diff --git a/src/main/java/org/apache/commons/dbcp2/PoolingDataSource.java b/src/main/java/org/apache/commons/dbcp2/PoolingDataSource.java
index 319c36cf90..52f69ba12e 100644
--- a/src/main/java/org/apache/commons/dbcp2/PoolingDataSource.java
+++ b/src/main/java/org/apache/commons/dbcp2/PoolingDataSource.java
@@ -143,6 +143,10 @@ public Connection getConnection() throws SQLException {
             throw new SQLException("Cannot get a connection, pool error " + e.getMessage(), e);
         } catch(final RuntimeException e) {
             throw e;
+        } catch(final InterruptedException e) {
+            // Reset the interrupt status so it is visible to callers
+            Thread.currentThread().interrupt();
+            throw new SQLException("Cannot get a connection, general error", e);
         } catch(final Exception e) {
             throw new SQLException("Cannot get a connection, general error", e);
         }
